File: github/typescript/src/services/organizations.ts
/**
 * GitHub Organizations Service
 * Provides access to GitHub organizations, members, and teams
 */

export interface RequestOptions {
  timeout?: number;
  headers?: Record<string, string>;
}

export interface ResilienceOrchestrator {
  request<T>(options: RequestConfig): Promise<T>;
}

export interface RequestConfig {
  method: string;
  path: string;
  timeout?: number;
  headers?: Record<string, string>;
  body?: unknown;
  query?: Record<string, string>;
}

// Organization types
export interface Organization {
  login: string;
  id: number;
  node_id: string;
  url: string;
  repos_url: string;
  events_url: string;
  hooks_url: string;
  issues_url: string;
  members_url: string;
  public_members_url: string;
  avatar_url: string;
  description?: string;
  name?: string;
  company?: string;
  blog?: string;
  location?: string;
  email?: string;
  twitter_username?: string;
  is_verified?: boolean;
  has_organization_projects?: boolean;
  has_repository_projects?: boolean;
  public_repos?: number;
  public_gists?: number;
  followers?: number;
  following?: number;
  html_url?: string;
  created_at?: string;
  updated_at?: string;
  type?: string;
}

export interface UpdateOrganizationRequest {
  billing_email?: string;
  company?: string;
  email?: string;
  twitter_username?: string;
  location?: string;
  name?: string;
  description?: string;
  has_organization_projects?: boolean;
  has_repository_projects?: boolean;
  default_repository_permission?: 'read' | 'write' | 'admin' | 'none';
  members_can_create_repositories?: boolean;
  members_can_create_internal_repositories?: boolean;
  members_can_create_private_repositories?: boolean;
  members_can_create_public_repositories?: boolean;
  members_allowed_repository_creation_type?: 'all' | 'private' | 'none';
  members_can_create_pages?: boolean;
  members_can_fork_private_repositories?: boolean;
  blog?: string;
}

export interface ListOrganizationsParams {
  since?: number;
  per_page?: number;
}

// Member types
export interface Member {
  login: string;
  id: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  url: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  type: string;
  site_admin: boolean;
}

export interface Membership {
  url: string;
  state: 'active' | 'pending';
  role: 'admin' | 'member';
  organization_url: string;
  organization: Organization;
  user: Member;
}

export interface UpdateMembershipRequest {
  role?: 'admin' | 'member';
}

export interface ListMembersParams {
  filter?: 'all' | '2fa_disabled';
  role?: 'all' | 'admin' | 'member';
  per_page?: number;
  page?: number;
}

// Team types
export interface Team {
  id: number;
  node_id: string;
  url: string;
  html_url: string;
  name: string;
  slug: string;
  description?: string;
  privacy?: 'secret' | 'closed';
  permission: 'pull' | 'push' | 'admin' | 'maintain' | 'triage';
  members_url: string;
  repositories_url: string;
  parent?: Team;
}

export interface CreateTeamRequest {
  name: string;
  description?: string;
  maintainers?: string[];
  repo_names?: string[];
  privacy?: 'secret' | 'closed';
  permission?: 'pull' | 'push' | 'admin' | 'maintain' | 'triage';
  parent_team_id?: number;
}

export interface UpdateTeamRequest {
  name?: string;
  description?: string;
  privacy?: 'secret' | 'closed';
  permission?: 'pull' | 'push' | 'admin' | 'maintain' | 'triage';
  parent_team_id?: number;
}

export interface ListTeamsParams {
  per_page?: number;
  page?: number;
}

export interface TeamMembership {
  url: string;
  role: 'member' | 'maintainer';
  state: 'active' | 'pending';
}

export interface AddTeamMemberRequest {
  role?: 'member' | 'maintainer';
}

/**
 * Organizations Service
 */
export class OrganizationsService {
  constructor(private readonly orchestrator: ResilienceOrchestrator) {}

  // Organization methods
  async list(params?: ListOrganizationsParams, options?: RequestOptions): Promise<Organization[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.since) query.since = params.since.toString();
      if (params.per_page) query.per_page = params.per_page.toString();
    }

    return this.orchestrator.request<Organization[]>({
      method: 'GET',
      path: '/organizations',
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async listForUser(username: string, params?: ListOrganizationsParams, options?: RequestOptions): Promise<Organization[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.per_page) query.per_page = params.per_page.toString();
    }

    return this.orchestrator.request<Organization[]>({
      method: 'GET',
      path: `/users/${encodeURIComponent(username)}/orgs`,
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async listForAuthenticatedUser(params?: ListOrganizationsParams, options?: RequestOptions): Promise<Organization[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.per_page) query.per_page = params.per_page.toString();
    }

    return this.orchestrator.request<Organization[]>({
      method: 'GET',
      path: '/user/orgs',
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async get(org: string, options?: RequestOptions): Promise<Organization> {
    return this.orchestrator.request<Organization>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}`,
      ...options,
    });
  }

  async update(org: string, request: UpdateOrganizationRequest, options?: RequestOptions): Promise<Organization> {
    return this.orchestrator.request<Organization>({
      method: 'PATCH',
      path: `/orgs/${encodeURIComponent(org)}`,
      body: request,
      ...options,
    });
  }

  // Member methods
  async listMembers(org: string, params?: ListMembersParams, options?: RequestOptions): Promise<Member[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.filter) query.filter = params.filter;
      if (params.role) query.role = params.role;
      if (params.per_page) query.per_page = params.per_page.toString();
      if (params.page) query.page = params.page.toString();
    }

    return this.orchestrator.request<Member[]>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/members`,
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async getMembership(org: string, username: string, options?: RequestOptions): Promise<Membership> {
    return this.orchestrator.request<Membership>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/memberships/${encodeURIComponent(username)}`,
      ...options,
    });
  }

  async updateMembership(
    org: string,
    username: string,
    request: UpdateMembershipRequest,
    options?: RequestOptions
  ): Promise<Membership> {
    return this.orchestrator.request<Membership>({
      method: 'PUT',
      path: `/orgs/${encodeURIComponent(org)}/memberships/${encodeURIComponent(username)}`,
      body: request,
      ...options,
    });
  }

  async removeMember(org: string, username: string, options?: RequestOptions): Promise<void> {
    await this.orchestrator.request<void>({
      method: 'DELETE',
      path: `/orgs/${encodeURIComponent(org)}/members/${encodeURIComponent(username)}`,
      ...options,
    });
  }

  // Team methods
  async listTeams(org: string, params?: ListTeamsParams, options?: RequestOptions): Promise<Team[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.per_page) query.per_page = params.per_page.toString();
      if (params.page) query.page = params.page.toString();
    }

    return this.orchestrator.request<Team[]>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/teams`,
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async getTeam(org: string, teamSlug: string, options?: RequestOptions): Promise<Team> {
    return this.orchestrator.request<Team>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}`,
      ...options,
    });
  }

  async createTeam(org: string, request: CreateTeamRequest, options?: RequestOptions): Promise<Team> {
    return this.orchestrator.request<Team>({
      method: 'POST',
      path: `/orgs/${encodeURIComponent(org)}/teams`,
      body: request,
      ...options,
    });
  }

  async updateTeam(org: string, teamSlug: string, request: UpdateTeamRequest, options?: RequestOptions): Promise<Team> {
    return this.orchestrator.request<Team>({
      method: 'PATCH',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}`,
      body: request,
      ...options,
    });
  }

  async deleteTeam(org: string, teamSlug: string, options?: RequestOptions): Promise<void> {
    await this.orchestrator.request<void>({
      method: 'DELETE',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}`,
      ...options,
    });
  }

  // Team member methods
  async listTeamMembers(org: string, teamSlug: string, params?: ListTeamsParams, options?: RequestOptions): Promise<Member[]> {
    const query: Record<string, string> = {};
    if (params) {
      if (params.per_page) query.per_page = params.per_page.toString();
      if (params.page) query.page = params.page.toString();
    }

    return this.orchestrator.request<Member[]>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}/members`,
      query: Object.keys(query).length > 0 ? query : undefined,
      ...options,
    });
  }

  async getTeamMembership(
    org: string,
    teamSlug: string,
    username: string,
    options?: RequestOptions
  ): Promise<TeamMembership> {
    return this.orchestrator.request<TeamMembership>({
      method: 'GET',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}/memberships/${encodeURIComponent(username)}`,
      ...options,
    });
  }

  async addTeamMember(
    org: string,
    teamSlug: string,
    username: string,
    request?: AddTeamMemberRequest,
    options?: RequestOptions
  ): Promise<TeamMembership> {
    return this.orchestrator.request<TeamMembership>({
      method: 'PUT',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}/memberships/${encodeURIComponent(username)}`,
      body: request || {},
      ...options,
    });
  }

  async removeTeamMember(org: string, teamSlug: string, username: string, options?: RequestOptions): Promise<void> {
    await this.orchestrator.request<void>({
      method: 'DELETE',
      path: `/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}/memberships/${encodeURIComponent(username)}`,
      ...options,
    });
  }
}

File: github/rust/src/services/organizations.rs
//! Organization operations.

use crate::client::GitHubClient;
use crate::errors::GitHubResult;
use crate::types::{Organization, Repository, Team, TeamPrivacy, User};
use serde::{Deserialize, Serialize};

/// Service for organization operations.
pub struct OrganizationsService<'a> {
    client: &'a GitHubClient,
}

impl<'a> OrganizationsService<'a> {
    /// Creates a new organizations service.
    pub fn new(client: &'a GitHubClient) -> Self {
        Self { client }
    }

    /// Lists organizations for the authenticated user.
    pub async fn list(&self) -> GitHubResult<Vec<Organization>> {
        self.client.get("/user/orgs").await
    }

    /// Gets an organization.
    pub async fn get(&self, org: &str) -> GitHubResult<Organization> {
        self.client.get(&format!("/orgs/{}", org)).await
    }

    /// Updates an organization.
    pub async fn update(
        &self,
        org: &str,
        request: &UpdateOrgRequest,
    ) -> GitHubResult<Organization> {
        self.client.patch(&format!("/orgs/{}", org), request).await
    }

    // Members

    /// Lists members of an organization.
    pub async fn list_members(&self, org: &str) -> GitHubResult<Vec<User>> {
        self.client.get(&format!("/orgs/{}/members", org)).await
    }

    /// Checks if a user is a member.
    pub async fn check_membership(&self, org: &str, username: &str) -> GitHubResult<bool> {
        let response = self
            .client
            .raw_request(
                reqwest::Method::GET,
                &format!("/orgs/{}/members/{}", org, username),
                Option::<()>::None,
            )
            .await;

        match response {
            Ok(_) => Ok(true),
            Err(e) if e.status_code() == Some(404) => Ok(false),
            Err(e) => Err(e),
        }
    }

    /// Gets membership for a user.
    pub async fn get_membership(&self, org: &str, username: &str) -> GitHubResult<Membership> {
        self.client
            .get(&format!("/orgs/{}/memberships/{}", org, username))
            .await
    }

    /// Sets membership for a user.
    pub async fn set_membership(
        &self,
        org: &str,
        username: &str,
        role: MembershipRole,
    ) -> GitHubResult<Membership> {
        let request = SetMembershipRequest { role };
        self.client
            .put(&format!("/orgs/{}/memberships/{}", org, username), &request)
            .await
    }

    /// Removes a member from an organization.
    pub async fn remove_member(&self, org: &str, username: &str) -> GitHubResult<()> {
        self.client
            .delete(&format!("/orgs/{}/members/{}", org, username))
            .await
    }

    // Teams

    /// Lists teams in an organization.
    pub async fn list_teams(&self, org: &str) -> GitHubResult<Vec<Team>> {
        self.client.get(&format!("/orgs/{}/teams", org)).await
    }

    /// Gets a team by slug.
    pub async fn get_team(&self, org: &str, team_slug: &str) -> GitHubResult<Team> {
        self.client
            .get(&format!("/orgs/{}/teams/{}", org, team_slug))
            .await
    }

    /// Creates a team.
    pub async fn create_team(&self, org: &str, request: &CreateTeamRequest) -> GitHubResult<Team> {
        self.client
            .post(&format!("/orgs/{}/teams", org), request)
            .await
    }

    /// Updates a team.
    pub async fn update_team(
        &self,
        org: &str,
        team_slug: &str,
        request: &UpdateTeamRequest,
    ) -> GitHubResult<Team> {
        self.client
            .patch(&format!("/orgs/{}/teams/{}", org, team_slug), request)
            .await
    }

    /// Deletes a team.
    pub async fn delete_team(&self, org: &str, team_slug: &str) -> GitHubResult<()> {
        self.client
            .delete(&format!("/orgs/{}/teams/{}", org, team_slug))
            .await
    }

    // Team Members

    /// Lists members of a team.
    pub async fn list_team_members(&self, org: &str, team_slug: &str) -> GitHubResult<Vec<User>> {
        self.client
            .get(&format!("/orgs/{}/teams/{}/members", org, team_slug))
            .await
    }

    /// Gets team membership for a user.
    pub async fn get_team_membership(
        &self,
        org: &str,
        team_slug: &str,
        username: &str,
    ) -> GitHubResult<TeamMembership> {
        self.client
            .get(&format!(
                "/orgs/{}/teams/{}/memberships/{}",
                org, team_slug, username
            ))
            .await
    }

    /// Adds or updates team membership for a user.
    pub async fn add_team_member(
        &self,
        org: &str,
        team_slug: &str,
        username: &str,
        role: TeamMemberRole,
    ) -> GitHubResult<TeamMembership> {
        let request = AddTeamMemberRequest { role };
        self.client
            .put(
                &format!("/orgs/{}/teams/{}/memberships/{}", org, team_slug, username),
                &request,
            )
            .await
    }

    /// Removes a user from a team.
    pub async fn remove_team_member(
        &self,
        org: &str,
        team_slug: &str,
        username: &str,
    ) -> GitHubResult<()> {
        self.client
            .delete(&format!(
                "/orgs/{}/teams/{}/memberships/{}",
                org, team_slug, username
            ))
            .await
    }

    // Team Repos

    /// Lists repositories for a team.
    pub async fn list_team_repos(
        &self,
        org: &str,
        team_slug: &str,
    ) -> GitHubResult<Vec<Repository>> {
        self.client
            .get(&format!("/orgs/{}/teams/{}/repos", org, team_slug))
            .await
    }

    /// Adds a repository to a team.
    pub async fn add_team_repo(
        &self,
        org: &str,
        team_slug: &str,
        owner: &str,
        repo: &str,
        permission: TeamRepoPermission,
    ) -> GitHubResult<()> {
        let request = AddTeamRepoRequest { permission };
        self.client
            .put_no_response(
                &format!("/orgs/{}/teams/{}/repos/{}/{}", org, team_slug, owner, repo),
                &request,
            )
            .await
    }

    /// Removes a repository from a team.
    pub async fn remove_team_repo(
        &self,
        org: &str,
        team_slug: &str,
        owner: &str,
        repo: &str,
    ) -> GitHubResult<()> {
        self.client
            .delete(&format!(
                "/orgs/{}/teams/{}/repos/{}/{}",
                org, team_slug, owner, repo
            ))
            .await
    }
}

/// Request to update an organization.
#[derive(Debug, Clone, Default, Serialize)]
pub struct UpdateOrgRequest {
    /// Billing email.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_email: Option<String>,
    /// Company.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub company: Option<String>,
    /// Email.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    /// Twitter username.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub twitter_username: Option<String>,
    /// Location.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<String>,
    /// Name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Blog URL.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blog: Option<String>,
}

/// Organization membership.
#[derive(Debug, Clone, Deserialize)]
pub struct Membership {
    /// State.
    pub state: MembershipState,
    /// Role.
    pub role: MembershipRole,
    /// Organization URL.
    pub organization_url: String,
    /// User.
    pub user: User,
    /// Organization.
    pub organization: Organization,
}

/// Membership state.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MembershipState {
    Active,
    Pending,
}

/// Membership role.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MembershipRole {
    Admin,
    Member,
}

#[derive(Debug, Clone, Serialize)]
struct SetMembershipRequest {
    role: MembershipRole,
}

/// Request to create a team.
#[derive(Debug, Clone, Serialize)]
pub struct CreateTeamRequest {
    /// Team name.
    pub name: String,
    /// Team description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Maintainers (usernames).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub maintainers: Option<Vec<String>>,
    /// Repository names.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub repo_names: Option<Vec<String>>,
    /// Privacy level.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub privacy: Option<TeamPrivacy>,
    /// Parent team ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_team_id: Option<u64>,
}

/// Request to update a team.
#[derive(Debug, Clone, Default, Serialize)]
pub struct UpdateTeamRequest {
    /// Team name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Team description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Privacy level.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub privacy: Option<TeamPrivacy>,
    /// Parent team ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_team_id: Option<u64>,
}

/// Team membership.
#[derive(Debug, Clone, Deserialize)]
pub struct TeamMembership {
    /// State.
    pub state: MembershipState,
    /// Role.
    pub role: TeamMemberRole,
    /// URL.
    pub url: String,
}

/// Team member role.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum TeamMemberRole {
    Member,
    Maintainer,
}

#[derive(Debug, Clone, Serialize)]
struct AddTeamMemberRequest {
    role: TeamMemberRole,
}

/// Team repository permission.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum TeamRepoPermission {
    Pull,
    Push,
    Admin,
    Maintain,
    Triage,
}

#[derive(Debug, Clone, Serialize)]
struct AddTeamRepoRequest {
    permission: TeamRepoPermission,
}
